File: Progs/libMax/maxbot-go/context.go
package maxbot

import (
	"context"
	"strings"
)

type Context struct {
	ctx    context.Context
	Client *Client
	Update Update
}

func (c *Context) Context() context.Context {
	return c.ctx
}

func (c *Context) Message() *Message {
	return c.Update.Message
}

func (c *Context) Callback() *CallbackQuery {
	return c.Update.Callback
}

func (c *Context) HasMessage() bool {
	return c.Update.Message != nil
}

func (c *Context) HasCallback() bool {
	return c.Update.Callback != nil
}

func (c *Context) MessageText() string {
	if c.Update.Message == nil {
		return ""
	}
	return strings.TrimSpace(c.Update.Message.Text)
}

func (c *Context) CallbackData() string {
	if c.Update.Callback == nil {
		return ""
	}
	return strings.TrimSpace(c.Update.Callback.Data)
}

func (c *Context) Command() string {
	if c.Update.Message == nil {
		return ""
	}
	return c.Update.Message.Command()
}

func (c *Context) IsCommand(cmd string) bool {
	cmd = strings.TrimPrefix(strings.TrimSpace(strings.ToLower(cmd)), "/")
	if cmd == "" {
		return false
	}
	return c.Command() == cmd
}

func (c *Context) ChatID() ID {
	if c.Update.Message != nil {
		return c.Update.Message.Chat.ID
	}
	if c.Update.Callback != nil {
		if c.Update.Callback.Chat != nil {
			return c.Update.Callback.Chat.ID
		}
		if c.Update.Callback.Msg != nil {
			return c.Update.Callback.Msg.Chat.ID
		}
	}
	return ""
}

func (c *Context) Reply(text string) error {
	chatID := c.ChatID()
	if chatID == "" {
		return nil
	}
	return c.Client.SendMessage(c.ctx, SendMessageRequest{
		ChatID: chatID,
		Text:   text,
	})
}

File: Progs/libMax/maxbot-js/src/context.ts
import type { CallbackQuery, ID, Message, Update } from './types';
import type { Client } from './client';

export class Context {
  readonly client: Client;
  readonly update: Update;

  constructor(client: Client, update: Update) {
    this.client = client;
    this.update = update;
  }

  message(): Message | undefined {
    return this.update.message;
  }

  callback(): CallbackQuery | undefined {
    return this.update.callback_query;
  }

  hasMessage(): boolean {
    return Boolean(this.update.message);
  }

  hasCallback(): boolean {
    return Boolean(this.update.callback_query);
  }

  messageText(): string {
    return this.update.message?.text?.trim() ?? '';
  }

  callbackData(): string {
    return this.update.callback_query?.data?.trim() ?? '';
  }

  command(): string {
    const text = this.messageText();
    if (!text.startsWith('/')) return '';
    const firstToken = text.slice(1).trim().split(/\s+/)[0] ?? '';
    const cmd = firstToken.includes('@') ? firstToken.split('@')[0] : firstToken;
    return cmd.trim().toLowerCase();
  }

  isCommand(cmd: string): boolean {
    const normalized = cmd.trim().replace(/^\//, '').toLowerCase();
    if (!normalized) return false;
    return this.command() === normalized;
  }

  chatID(): ID | '' {
    return this.update.message?.chat.chat_id ?? this.update.callback_query?.chat?.chat_id ?? this.update.callback_query?.message?.chat.chat_id ?? '';
  }

  async reply(text: string, signal?: AbortSignal): Promise<void> {
    const chatID = this.chatID();
    if (!chatID) return;
    await this.client.sendMessage({ chat_id: chatID, text }, signal);
  }
}
